### PR TITLE
advanced label formatting,multiline feature and labelBackground

### PR DESCRIPTION
**Pull request by hjhimanshu01**
Feature and how to access:

1. Default alignment is set to "left" and newline is given at each whitespace (according to \s regex) 
2. for negative half of the plane in y direction , height of label is dynamically changed so as to fit in 
    canvas
3. for positive half of the plane in y direction , label is made to start at almost the center of the label
4. for x axis , a suitable width (width of longest word in the sentence is selected) and that width is used to 
    given spacing in positive x while in negative x axis , it starts near to center of node.

**Updated : label formatting features added to the library to change font size , weight and family on the fly and optimized distance transform calculation**
_**Labels Background feature added ( needs more polishing and borders around labelsBackground as a further task )**_
```
node: {
            minSize: 6,
            maxSize: 16,
            color: "rgb(255, 0, 0)",
            texture: "images/circle.png",
            label: {
              color: "rgb(120, 120, 120)",
              size: 14,
              font: {
                strokeText: false,
                size: 30,
                family: "vedana",
                weight: "normal",
                alignment: "center",
                type: "sdf",
              }
```

ccNetViz-master

@saskaale , reviews please?

### CODE REVIEW

**Comment by saskaale** [comment]
Hello @brickmaker ,
thank you for your changes. Can you please modify the API so the configuration of the aligment would be inside the font object? 

node: {
            minSize: 6,
            maxSize: 16,
            color: "rgb(255, 0, 0)",
            texture: "images/circle.png",
            label: {
              color: "rgb(120, 120, 120)",
              size: 14,
              font: {
                align: "middle",
                type: "sdf",
              }
}

I would also completely leave out the second option through the direct setting in the nodes.


Also speaking about the conentions, it would be nice if you would follow the CSS standart so the options would be the "left|right|center".

thanks,
Ales

File: src/texts/sdf/sdf.js [hunk 1]
```diff
@@ -117,36 +117,177 @@ export default class {
...
   }
+  
+  // function to align text left, right or center
+  alignText(alignment, text, x, y, width, height, markDirty, widthArray, wordWidth) {
 
+    // x and y are the clipspace co-ordinates between 0 and 1
+   // dx and dy shifts the position of label w.r.t possibly node
+   // (TODO: dx and dy are calculated w.r.t what is not clear , please clear it if you find out)
+    
+   if(!( alignment === 'right' || alignment === 'left' || alignment === 'middle')) {
```
Comment (saskaale) [inline src/texts/sdf/sdf.js]: please use the == instead

File: src/texts/sdf/sdf.js [hunk 2]
```diff
@@ -117,36 +117,177 @@ export default class {
...
+            dx = x <= 0.5 ? 0 : -width ;
+            dy = dy-Math.floor(height/3);
+          } else {
+            const char = this._getChar(text[i], markDirty);
+            const rect = char.rect || {};
+            let horiBearingX = 3;
+            dx += horiBearingX;
+     
+            // rect.x rect.w rect.h rect.y are all atlas widths heigths x y positions etc 
+            ret.push({
```
Comment (saskaale) [inline src/texts/sdf/sdf.js]: this part looks like a lot of the duplication of code. Is it possible wrap this part of the code as a function?

Comment (hjhimanshu01) [reply src/texts/sdf/sdf.js]: Sure, did it!

File: src/texts/sdf/sdf.js (hunk 1 above)
Comment (hjhimanshu01) [reply src/texts/sdf/sdf.js]: Resolved!

**Comment by hjhimanshu01** [comment]
@saskaale , updated the fontSize feature and labelBackground now works pretty well, also fontSize now denotes basically labelSize, that improves the current clustered sdf.html example

**Pull request closed without merging**